app/controllers/tags_controller.rb

```diff
@@ -117,6 +117,35 @@ def update
     end
   end
 
+  def upload
+    guardian.ensure_can_admin_tags!
+
+    file = params[:file] || params[:files].first
+
+    hijack do
+      begin
+        Tag.transaction do
+          CSV.foreach(file.tempfile) do |row|
+            raise Discourse::InvalidParameters.new(I18n.t("tags.upload_row_too_long")) if row.length > 2
+
+            tag_name = DiscourseTagging.clean_tag(row[0])
+            tag_group_name = row[1] || nil
+
+            tag = Tag.find_by_name(tag_name) || Tag.create!(name: tag_name)
+
+            if tag_group_name
+              tag_group = TagGroup.find_by(name: tag_group_name) || TagGroup.create!(name: tag_group_name)
```

Comment:
I think we can use `find_or_create_by_safe!` here

Reply:
For tags, we can't because we need to use the case insensitive `find_by_name` function. For tag groups, we can, but I think it would make the code a little more confusing to read since it will be different to tags. What do you think @tgxworld?

Reply:
I guess this is fine too since `find_or_create_by` is really just a helper method that does the same thing.